ASoC: SOF: ipc4-loader/topology: Query the module's CPC from firmware manifest only

ujfalusi opened the pull request:
Hi,

This is a scaled back version of #4315 only implementing the CPC querying from the firmware manifest's module configuration section.

The manifest contains a module config section which contains measured CPC of a module depending on it's configuration (IBS/OBS). Use this information in MOD_INIT_INSTANCE message to send precise information to the firmware.

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -423,6 +423,65 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
+	int i;
+
+	if (!fw_module || !fw_module->fw_mod_cfg)
+		return;
```
is this not an error?

It'd be worth a comment to explain when this branch is taken and why

ujfalusi replied:
This should not happen, I have it here for safety.
We also have it in other places.

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -423,6 +423,65 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
+
+	/*
+	 * Find the best matching (highest) CPC value based on the module's
+	 * external configuration (IBS/OBS).
```
what did you mean by 'external'? Did you mean "the module IBS/OBS configuration retrieved from the topology"?

ujfalusi replied:
The IBS/OBS is in mod_cfg along with the CPC as triplet.
The IBS/OBS for the used configuration is coming from topology when the module is configured for the stream.
IBS/OBS is the 'external' configuration in a sense that it is the input and output buffer size, which is related to the audio formats picked for the given side of the module.

plbossart replied:
@ujfalusi I am in agreement with your explanation, but still 'external' is very ambiguous.

Would this work for you

"
Find the best matching (highest) CPC value based on the module's IBS/OBS configuration inferred from the audio format selection.
"

lgirdwood replied:
"external configuration" -> "metadata found in the FW binary file".

ujfalusi replied:
@lgirdwood, @plbossart, yes, I forgot to update this, in the next version I will correct it.

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -423,6 +423,65 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
+	 * encountered while parsing the module config entries
+	 */
+	if (!basecfg->cpc)
+		basecfg->cpc = max_cpc;
```
I feel this is an error: the topology uses an ibs/obs value that is not aligned with the manifest. In that case, using the maximum cpc available is just inventing information. it's not reliable and it's not recommended.

I would just reject this configuration completely.

And see below why it's inconsistent ...

ujfalusi replied:
We cannot reject this as nothing will work if a module has no matching mod_cfg entry, we print an info and that will give us the information what needs to be added to manifest.

plbossart replied:
what makes you think the max is relevant or even works @ujfalusi ?

If for example we want to deal with 192kHz data, and somehow this leads to an IBS/OBS pair that is not listed in the manifest, it's possible that the max listed in the manifest is a low estimate of the actual needs.

So there's a strong likelihood that we would be trying our luck and fail later with random xruns.

My approach is always: fail big and fail early. Trying to make things work by accident is not a reliable solution.

Or at the very least add some linear fudge factor based on the ratio between IBS/OBS listed and IBS/OBS requested.

ujfalusi replied:
OK, so treat the no exact match as no CPC value at all and use 0 (with a print)?

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -966,6 +966,14 @@ sof_ipc4_update_resource_usage(struct snd_sof_dev *sdev, struct snd_sof_widget *
+
+	dev_dbg(sdev->dev, "%s: ibs / obs / cpc: %u / %u / %u\n",
+		swidget->widget->name, base_config->ibs, base_config->obs,
+		base_config->cpc);
```
... so here are logging information that is inconsistent because there is no matching ibs/obs/cpc triplet.

ujfalusi replied:
Line 478 prints the information if there is no matching IBS/OBS was found, but we did ended up with a non 0 CPC.
If you have deb level enabled it will look like this:
```
COPIER (UUID: 9BA00C83-CA12-4A83-943C-1FA2E82F9DDA): No CPC match for ibs/obs: 192/384
copier.host.1.1: ibs / obs / cpc: 192 / 384 / 7272
```
It is not relevant in the dev_info() print what is the CPC we ended up using, what it matters is that we have no match.

If there is no match, we will print with dev_warn() first:
```
COPIER (UUID: 9BA00C83-CA12-4A83-943C-1FA2E82F9DDA): No CPC value available
copier.host.1.1: ibs / obs / cpc: 192 / 384 / 0
```

plbossart replied:
@ujfalusi I am not clear why there are three levels of information with warn, info and dbg. Two of them are errors in my book and we should stop right there. There is an invalid configuration, we stop. end of story.

ujfalusi replied:
and have no audio card at all?
It might take 4 months for a user to get a new firmware with updated manifest..

What if someone experimenting with topology files and starts to use a module which does not happen to have the exact same config measured? We break audio?

If we do that with this PR than the next step is a revert of the series as it introduces a regression for users.

I don't think that is something we can do and should do.

All red cavs and avs PR test would be granted.

ujfalusi replied:
If you don't have dev_dbg() enabled then the information about the use ibs/obs/cpc is not shown. User will see either a warning that there were no CPC at all (the ibs/obs is irrelevant as there is not a single one can be found in mod_cfg) or an info print that while there are CPC values, the ibs/obs pair has no matching hit.
If there is a match, nothing is printed.

And if I think about this, then likely we need to warn if there is no mod_cfg either for the module from where we would be taking the CPC.

plbossart replied:
We are asked to parse the manifest and provide the CPC value for a given IBS/OBS. If the CPC doesn't exist for this configuration, well end of story. The driver should not invent information that should have been provided by the firmware integrator.

If someone tinkers with the topology file and breaks the CPC mapping, it's a revert of the topology files. very simple. If they ship a broken topology without testing that it is compatible with the manifest, there's nothing we can do.

We have lots of cases like this today, if someone changes the topology and refers to a non-existing firmware component, then we just don't have a card.

We just don't have ways to deal with bad topology configurations other than failing to create a card.

ujfalusi replied:
So, you want to fail with error and not allow the playback if we  don't have CPC with direct match from the manifest?
Either if  the mod_cfg does not have a matching entry or if there is no mod_cfg at all?

If we do this than you are not going to be able to bisect firmware beyond the point the manifest has been updated as the kernel will refuse to work.

plbossart replied:
if the kernel worked in the past and suddendly refuses to work, that would be because of a topology change that modified IBS/OBS, or an update of the manifest. At that point, the changed configuration will be detected as failing CI tests and prevent the update from being merged. 

If we don't fail, we will let a bad configuration be merged and no one will ever attempt to fix it. It's the responsibility of the firmware team to make sure the manifest makes sense and covers all allowed usages.

plbossart reviewed (changes requested):
Couple of nit-picks on comments, and one push-back on the selection of the highest cpc value if there is no IPS/OBS match.

ujfalusi commented:
Changes since v1:
- Only allow direct match for CPC (based on IBS/OBS matching)

If there is no mod_cfg or all CPC values are 0 or we don't have direct patch we will fail and print relevant error message.

Note: _all_ cavs and avs test cases will fail which tries to use audio as the firmware's manifest does not have correct mod_cfg.

ujfalusi commented:
Changes since v2:
- comment updated for the CPC lookup

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -423,6 +423,75 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
+	if (!max_cpc)
+		msg = "No CPC value available in firmware file metadata";
+	else if (!cpc_pick)
+		msg = "No CPC match in firmware file metadata";
```
nit-pick: I don't know what metadata means. "firmware file manifest' seems more obvious to me.

ujfalusi replied:
I would be happy to change it back to manifest...

ujfalusi replied:
Ah, you mean to put the `firmware file manifest` in quotation?

plbossart replied:
yes, but that's fine as is. let's stop wordsmithing.

plbossart reviewed (approved):
Thanks @ujfalusi  - one optional wording suggestion below.

ranj063 commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -423,6 +423,75 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
+		&fw_module->man4_module_entry.uuid, msg, basecfg->ibs,
+		basecfg->obs);
+	dev_err(sdev->dev, "Please try to update the firmware. If the issue persists,\n");
+	dev_err(sdev->dev, "file a bug at https://github.com/thesofproject/sof/issues/\n");
```
oops didnt see the the sentence below. ignore my previous comment please

ranj063 commented:
wow, failures all over @ujfalusi

plbossart commented:
> wow, failures all over @ujfalusi

well d'uh, the firmware manifests don't contain the expected CPC...

ranj063 commented:
> > wow, failures all over @ujfalusi
> 
> well d'uh, the firmware manifests don't contain the expected CPC...

@plbossart then we must fix our FW first before we merge this isnt it?

plbossart commented:
> > > wow, failures all over @ujfalusi
> > 
> > 
> > well d'uh, the firmware manifests don't contain the expected CPC...
> 
> @plbossart then we must fix our FW first before we merge this isnt it?

I think so. I suggested to @lgirdwood and @ujfalusi that we have a formal SOF issue tracking this manifest problem.

ujfalusi commented:
Changes since v3:
- Do not fail on missing CPC value from manifest, print warning and use 0 as CPC value

This aligns with how the firmware is going to handle the CPC: https://github.com/thesofproject/sof/pull/7575
If it receives 0 then it takes internal fallback value of maximum CPC for the DSP we are running.
The kernel prints warning asking user to report firmware issues so they can be fixed.

plbossart commented:
> Changes since v3:
> 
> * Do not fail on missing CPC value from manifest, print warning and use 0 as CPC value
> 
> This aligns with how the firmware is going to handle the CPC: [thesofproject/sof#7575](https://github.com/thesofproject/sof/pull/7575) If it receives 0 then it takes internal fallback value of maximum CPC for the DSP we are running. The kernel prints warning asking user to report firmware issues so they can be fixed.

The problem with dev_warn is that no one looks at them. sof-test will report a PASS status.

I would recommend we have a additional patch that makes it a dev_err() and we keep internal for CI purposes (i.e. not upstream it). Otherwise we will not keep track of the issue and nothing will happen.

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -423,6 +423,71 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
+		msg = "No CPC match in the firmware file's manifest";
+
+no_cpc:
+	dev_warn(sdev->dev, "%s (UUID: %pUL): %s (ibs/obs: %u/%u)\n",
```
the problem is that no one looks at warnings, because they are not clear.

In this case, the message makes zero sense for an average user who doesn't even know what the CPC is.

It should really be "Invalid configuration: <msg>, the firmware will force the DSP clock to run with fastest clock" or something more meaningful

And I would use dev_err() as a way to catch those errors in CI, we do not track dev_warn, ever.

ranj063 replied:
But @plbossart an error would mean we need https://github.com/thesofproject/sof/pull/7574 before we merge this

plbossart replied:
https://github.com/thesofproject/sof/pull/7575 was merged, it replaced  thesofproject/sof#7574 

we can add a dev_err() in a follow-up patch later, for our own CI tracking.

I still think the message is far from clear. Put yourself in the shoes of johnny 4-pack, how on earth could they know what CPC is. 

Maybe we can deal with this as a fixup, I'll approve to avoid blocking progress (or the new definition of progress).

ranj063 replied:
7575 only went into mtl-005-drop-stable. I dont see anything for main

plbossart reviewed (approved):
approving to unblock, but comment needs to be updated before upstream

plbossart commented:
SOFCI TEST
